docs/architecture/adr-006-upgrade-client-implementation.md

```diff
@@ -0,0 +1,433 @@
+their clients. However there are general rules that apply to all and can serve
+as a framework for any `IBC-rs` implementation. On this basis, this record aims
+to justify the logic behind upgrading a light client, determine the boundary
+between client-wide and client-specific upgrade processes, list requisites for
+validation and execution steps, and explain Tendermint's upgrade client
+implementation within the
+[ics07_tendermint](../../crates/ibc/src/clients/ics07_tendermint).
+
+## Decision
+
+In this section, after presenting rules that mainly derived from the IBC
+protocol and the review of the `IBC-Go` implementation, the upgrade client
+process for Tendermint chains is outlined in detail. Then, the design and
+implementation of this rationale in `IBC-rs` is explained.
+
+### Chain-wide Upgrade Rules
+
+#### Chain supports IBC upgrades?
+
+* IBC currently **ONLY** supports planned upgrades that are committed to in
+  advance by the upgrading chain, in order for counterparty clients to maintain
+  their connections securely.
+* An IBC upgrade **MUST** be performed if upgrading an IBC-connected chain
+  breaks the counterparty IBC client.
+* There **MUST** be a proof verification process to check upgraded client and
+  consensus states against the host chain's state.
+* Chain upgrades **MUST NOT** result in changing
+  [ClientState](../../crates/ibc/src/core/ics02_client/client_state.rs#ClientState)
+  or
+  [ConsensusState](../../crates/ibc/src/core/ics02_client/consensus_state.rs#ConsensusState)
+  implementations
+* It is **UP TO** the chain's architecture how updated client and consensus
+  states are committed, either through decentralized approaches, like governance
+  or centralized methods, like a multisig account, etc.
+* Upon commitment, chain's store **MUST** contain updated client and consensus
+  states, which can be retrieved using respective upgrade paths
+
+#### IBC Handler accepts upgrade?
+
+After ensuring that the chain upgrades is supported by IBC, the general
+validation and execution steps that apply to all chains are as follows. The
+criteria for classifying a validation as client-wide (also can be known as
+basic) or client-specific was whether that IBC handler can perform that action
+just using its own context data, which are available upon calling interfaces
+provided by `ClientState` and `ConsensusState` traits, like `is_frozen()`,
+`latest_height()`, etc.
+
+* Latest client state **MUST NOT** be frozen
+* Latest consensus state **MUST** be within the trusting period of the latest
+  client state
+* Received update message including verification proofs **MUST** successfully be
+  decoded into the domain types
+* Clients **MUST** ensure that the planned last height of the current revision
+  is somehow encoded in the proof verification process. This prevents premature
+  upgrades, as the counterparty may cancel or modify upgrade plans before the
+  last planned height.
```

Comment:
Check this [line](https://github.com/cosmos/ibc-go/blob/e650be91614ced7be687c30eb42714787a3bbc59/modules/core/exported/client.go#L117). We wont't send the message in advance. Actually, The purpose of this statement was to put an emphasize on "there must be a proof verification for the upgrade process" which occurs at the planned last height. Maybe I should reword it.

Reply:
If the upgrade process is planned to happen at height H, the counterparty client won’t accept a submitted upgrade message before height H - 1 (This is the last planned height at the current(old) version of chain). This is checked via proofs, as if they are not for the correct height (H - 1) the proof verifications would fail.
[d59ddfe](https://github.com/cosmos/ibc-rs/pull/383/commits/d59ddfe9c718ded5121c479e562338b9ebcc2c36)

Reply:
> This is checked via proofs, as if they are not for the correct height (H - 1) the proof verifications would fail.

I was going to ask you to point to me the line of code that does that. But I think I just understood what they meant by "the planned last height of the current revision is somehow encoded in the proof verification process". That didn't make sense to me, because all proofs in IBC are Merkle proofs, which doesn't check anything other than "this item is/isn't in this Merkle tree". 

So then how do you "encode the last height in the proof verification"? I think how they implement that is to only write to the store *at* the planned height, as part of the protocol. Therefore, if the Merkle proof verification passes, the added semantic is "we are currently at the last planned height".

So this is why I couldn't find any line of code that did that; it is implied by the protocol. So if I'm right here, it means that it is then a requirement for Tendermint *hosts* to only write the upgraded state *at the planned height* in order to be compatible with ibc-go's light client implementation.

@AdityaSripal can you confirm that this is correct?

Reply:
Exactly, My understanding was that.  It's implied by the protocol.

Reply:
Another thing we should note is:
The initial consensus state for the next chain is only committed at height H - 1 ([Ref](https://github.com/cosmos/ibc-go/blob/main/docs/ibc/upgrades/quick-guide.md#step-by-step-upgrade-process-for-sdk-chains)). So, there won't be any `UpgradedConsensusState` available at the upgrade path before that height and if a relayer tries to submit an upgrade message prior to that, it should fail.
I added a short description in this regard: https://github.com/cosmos/ibc-rs/pull/383/commits/7dcbdb48a860c6a5bd2efd03f079b3b1fcce1347